lib/src/rules/avoid_dynamic_calls.dart

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) 2021, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import '../analyzer.dart';
+
+const _desc = r'Avoid method calls and property access on a "dynamic" target.';
+
+const _details = r'''
+
+**DO** avoid method calls or property accessors on an object that is either
+explicitly or implicitly statically typed "dynamic". Dynamic calls are treated
+slightly different in every runtime environment and compiler, but most
+production modes (and even some development modes) have both compile size and
+runtime performance penalties associated with dynamic calls.
+
+Additionally, targets typed "dynamic" disabled most static analysis, meaning it
+is easier to lead to a runtime "NoSuchMethodError" or "NullError" than properly
+statically typed Dart code.
+
+Note, that despite "Function" being a type, the semantics are close to identical
+to "dynamic", and calls to an object that are typed "Function" will also trigger
+this lint.
+
+**BAD:**
+```
+void explicitDynamicType(dynamic object) {
+  print(object.foo());
+}
+
+void implicitDynamicType(object) {
+  print(object.foo());
+}
+
+abstract class SomeWrapper {
+  T doSomething<T>();
+}
+
+void inferredDynamicType(SomeWrapper wrapper) {
+  var object = wrapper.doSomething();
+  print(field.foo());
+}
+
+void functionType(Function function) {
+  function();
+}
+```
+
+**GOOD:**
+```
+void explicitType(Fooable object) {
+  object.foo();
+}
+
+void castedType(dynamic object) {
+  (object as Fooable).foo();
+}
+
+abstract class SomeWrapper {
+  T doSomething<T>();
+}
+
+void inferredType(SomeWrapper wrapper) {
+  var object = wrapper.doSomething<Fooable>();
+  object.foo();
+}
+```
+
+''';
+
+class AvoidDynamicCalls extends LintRule implements NodeLintRule {
+  AvoidDynamicCalls()
+      : super(
+          name: 'avoid_dynamic_calls',
+          description: _desc,
+          details: _details,
+          group: Group.errors,
```

Comment:
If this is experimental, then we should mark it as such by adding an argument of the form `maturity: Maturity.experimental`.